lib/httplib/csrf/csrf.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2016 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csrf
+
+import (
+	"crypto/subtle"
+	"encoding/hex"
+	"net/http"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+const (
+	// CookieName is a name of the cookie
+	CookieName = "grv_csrf"
+	// HeaderName is the default HTTP request header to inspect
+	HeaderName = "X-CSRF-Token"
+	// tokenLenBytes is CSRF token length in bytes.
+	tokenLenBytes = 32
+	// defaultMaxAge is the default MaxAge for cookies.
+	defaultMaxAge = 0
+)
+
+// AddCSRFProtection adds CSRF token into the user session via secure cookie,
+// it implements "double submit cookie" approach to check against csrf attacks
+// https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet#Double_Submit_Cookie
+func AddCSRFProtection(w http.ResponseWriter, r *http.Request) (string, error) {
+	encodedToken := ""
+	token, err := extractFromCookie(r)
+	// if there was an error retrieving the token, the token doesn't exist
+	if err != nil || len(token) != tokenLenBytes {
+		encodedToken, err = utils.CryptoRandomHex(tokenLenBytes)
```

Comment:
this logic was hard for me to read:

it turns out that this code generates crypto random token, both when `err !=nil and err == nil` but discards it in the former case.

I would rewrite it for clarity and efficiency:

```go
if err != nil {
     return trace.Wrap(err)
}
if len(token) != tokenLenBytes {
   return trace.BadParamter("unexpected token")
}
if len(token) == 0 {
   encodedToken, err = utils.CryptoRandomHex(tokenLenBytes)
   if err != nil {
       return "", trace.Wrap(err)
   }
} else {
    encodedToken = hex.EncodeToString(token)
}
```

Although from the code it is not clear why regenerate the token if tokenLenBytes does not match the token. It appears there are two cases: len(token) == 0 which is the case you want, and len(token) != tokenLenBytes that is unexpected and should be an error.

Reply:
This function is used by index.html handler which should always return a HTML. The reason to check for the length is make sure that a user still gets HTML (functioning app) if we decide to change the length of the token or encoder.  

I agree that this piece is not easy to follow but to me this is because  `utils.CryptoRandomHex(tokenLenBytes)` does both: generates a random value and does encoding.

Reply:
> This function is used by index.html handler which should always return a HTML. The reason to check for the length is make sure that a user still gets HTML (functioning app) if we decide to change the length of the token or encoder.

This code solves the problem we don't have now, therefore it is redundant.